# Notebook for running all the examples in the examples folder

This notebook also serves as a test of the ras_commander package using the HEC-RAS Project Examples provided by HEC.

Example project names can be found in the example_projects.csv file after the examples are run.

To use a different example, change the project name in the examples and run them directly or run this notebook.  


The notebook format provides a convenient way to run all the examples to ensure that no features are broken after code revisions.  


-----

This table is a valuable reference for formulating examples and anticipating script behavior.

| Project           | Type     | Existing Numbers | Next Available |
|-------------------|----------|-------------------|----------------|
| Balde Eagle Creek | Plan     | 01, 02            | 03             |
|                   | Flow     | 01, 02            | 03             |
|                   | Unsteady | 02                | 01, 03         |
|                   | Geometry | 01                | 02             |
| Muncie            | Plan     | 01, 03, 04        | 02, 05         |
|                   | Flow     | 01                | 02             |
|                   | Unsteady | 01                | 02             |
|                   | Geometry | 01, 02, 04        | 03, 05         |

Notes:
1. For Balde Eagle Creek's Unsteady entries, 01 is available before 03. The library should use the lowest available number when creating new entries.
2. For Muncie's Plan entries, both 02 and 05 are available. The library should use 02 first.
3. For Muncie's Geometry entries, 03 is available before 05.

When writing examples or scripts:
1. Use existing numbers when referencing current plans, flows, unsteady setups, or geometries.
2. When creating new entries (e.g., cloning), use the "Next Available" number and then increment for subsequent creations.
3. After operations that create new entries, verify that the ras object has been updated to reflect these changes.

This approach will help ensure that examples run as expected and that we're always working with valid, up-to-date data in our scripts.

-----

# RAS-Commander (ras_commander) library examples

### The Examples Below Have Been Tested and Verified.

-----

In [1]:
#!pip install pandas pathlib requests psutil h5py matplotlib scipy

In [2]:
# Example 1: Import and execute 01_project_initialization.py
print("Executing 01_project_initialization.py:")
%run 01_project_initialization.py
print("\n")

Executing 01_project_initialization.py:


2024-09-27 00:30:12,496 - INFO - Example projects folder: c:\GH\ras-commander\examples\example_projects
2024-09-27 00:30:12,497 - INFO - Found zip file: c:\GH\ras-commander\examples\Example_Projects_6_5.zip
2024-09-27 00:30:12,498 - INFO - Loading project data from CSV...
2024-09-27 00:30:12,503 - INFO - Loaded 66 projects from CSV. Use list_categories() and list_projects() to explore them.
2024-09-27 00:30:12,504 - INFO - ----- RasExamples Extracting Project -----
2024-09-27 00:30:12,505 - INFO - Extracting project 'Balde Eagle Creek'
2024-09-27 00:30:12,587 - INFO - Successfully extracted project 'Balde Eagle Creek' to c:\GH\ras-commander\examples\example_projects\Balde Eagle Creek
2024-09-27 00:30:12,587 - INFO - ----- RasExamples Extraction Complete -----
2024-09-27 00:30:12,588 - INFO - ----- RasExamples Extracting Project -----
2024-09-27 00:30:12,589 - INFO - Extracting project 'BaldEagleCrkMulti2D'
2024-09-27 00:30:12,590 - INFO - Project 'BaldEagleCrkMulti2D' already exists. D

Example Set 1: Using the default global 'ras' object
-----------------------------------------------------
Step 1: Initializing with global RAS instance

Global RAS Instance (Bald Eagle Creek) Project Data:
Project Name: BaldEagle
Project Folder: c:\GH\ras-commander\examples\example_projects\Balde Eagle Creek
PRJ File: C:\GH\ras-commander\examples\example_projects\Balde Eagle Creek\BaldEagle.prj
HEC-RAS Executable Path: C:\Program Files (x86)\HEC\HEC-RAS\6.5\Ras.exe

Plan Files DataFrame:
  plan_number                                          full_path  \
0          01  c:\GH\ras-commander\examples\example_projects\...   
1          02  c:\GH\ras-commander\examples\example_projects\...   

  computation_interval dss_file flow_file friction_slope_method geom_file  \
0                 2MIN      dss       u02                     2       g01   
1                 2MIN      dss       f02                     1       g01   

  mapping_interval                     plan_title program_version run

ValueError: No HEC-RAS project file found in c:\GH\ras-commander\examples\example_projects\Muncie

-----

In [3]:
import time

# wait 10 seconds
time.sleep(10)

In [ ]:
# Example 2: Import and execute 02_plan_operations.py
print("Executing 02_plan_operations.py:")
%run 02_plan_operations.py
print("\n")

-----

In [5]:
# wait 10 seconds
time.sleep(10)

In [ ]:
# Example 3: Import and execute 03_geometry_operations.py
print("Executing 03_geometry_operations.py:")
%run 03_geometry_operations.py
print("\n")

-----

In [7]:
# wait 10 seconds
time.sleep(10)

In [ ]:
# Example 4: Import and execute 04_unsteady_flow_operations.py
print("Executing 04_unsteady_flow_operations.py:")
%run 04_unsteady_flow_operations.py
print("\n")

-----

In [9]:
# wait 10 seconds
time.sleep(10)

In [ ]:
# Example 5: Import and execute 05_utility_functions.py
print("Executing 05_utility_functions.py:")
%run 05_utility_functions.py
print("\n")

-----

In [11]:
# wait 10 seconds
time.sleep(10)

In [ ]:
# Example 6: Import and execute 06_single_plan_execution.py
print("Executing 06_single_plan_execution.py:")
%run 06_single_plan_execution.py
print("\n")

-----

In [ ]:
# wait 10 seconds
time.sleep(10)

In [ ]:
# Example 7: Import and execute 07_sequential_plan_execution.py
print("Executing 07_sequential_plan_execution.py:")
%run 07_sequential_plan_execution.py
print("\n")

-----

In [ ]:
# wait 10 seconds
time.sleep(10)

In [ ]:
# Example 8: Import and execute 08_parallel_execution.py
print("Executing 08_parallel_execution.py:")
%run 08_parallel_execution.py
print("\n")

-----

In [ ]:
# wait 10 seconds
time.sleep(10)

In [ ]:
# Example 9: Import and execute 09_specifying_plans.py
print("Executing 09_specifying_plans.py:")
%run 09_specifying_plans.py
print("\n")

-----

In [ ]:
# wait 10 seconds
time.sleep(10)

In [ ]:
print("Executing 10_arguments_for_compute.py:")
%run 10_arguments_for_compute.py
print("\n")

In [ ]:
# wait 10 seconds
time.sleep(10)

In [ ]:
print("Executing 12_plan_set_execution.py:")
%run 12_plan_set_execution.py
print("\n")

In [ ]:
# wait 10 seconds
time.sleep(10)

In [ ]:
print("Executing 13_multiple_project_operations.py:")
%run 13_multiple_project_operations.py
print("\n")

In [ ]:
# wait 10 seconds
time.sleep(10)

In [ ]:
# Example 14 is long, and is a notebook.

In [ ]:
print("Executing 15_plan_key_operations.py:")
%run 15_plan_key_operations.py
print("\n")

In [ ]:
# wait 10 seconds
time.sleep(10)

In [ ]:
print("Executing 16_scanning_ras_project_info.py:")
%run 16_scanning_ras_project_info.py
print("\n")


In [ ]:
# wait 10 seconds
time.sleep(10)

In [ ]:
print("Executing xx_edge_cases.py:")
%run xx_edge_cases.py
print("\n")

# Guide for Locating Relevant Examples:

## Category: 1D Unsteady Flow Hydraulics

### Flow Hydrograph

| Example | Plan File |
|---------|-----------|
| Balde Eagle Creek | BaldEagle.p01, BaldEagle.p02 |
| ContractionExpansionMinorLosses | ContExp.p01, ContExp.p02, ContExp.p03, ContExp.p04 |
| Culvert Hydraulics | Beav_Culvert.p01 |
| Culverts with Flap Gates | CulvertFlap.p01 |
| Dam Breaching | BaldEagleDamBrk.p07, BaldEagleDamBrk.p02, BaldEagleDamBrk.p10, BaldEagleDamBrk.p06, BaldEagleDamBrk.p08, BaldEagleDamBrk.p04, BaldEagleDamBrk.p03, BaldEagleDamBrk.p05, BaldEagleDamBrk.p09, BaldEagleDamBrk.p11 |
| Elevation Controled Gates | ElevControledGates.p01 |
| Inline Structure with Gated Spillways | Inline_3Gates.p01 |
| Internal Stage and Flow Boundary Condition | IBStageFlowTest.p01 |
| JunctionHydraulics | JunctionHydraulics.p01, JunctionHydraulics.p02, JunctionHydraulics.p04, JunctionHydraulics.p03 |
| Lateral Strcuture with Gates | TwoGateGrps.p01 |
| Lateral Structure connected to a River Reach | 3Reach_with_Lat.p01 |
| Lateral Structure Overflow Weir | LatMultXS.p01 |
| Lateral Structure with Culverts | LatWeirCulverts.p01 |
| Lateral Structure with Culverts and Gates | Beav_Lateral.p01 |
| Levee Breaching | LeveeBreach.p01 |
| Multiple Reaches with Hydraulic Structures | 3ReachUnsteady.p01, 3ReachUnsteady.p02 |
| Pumping Station | Pumps.p01 |
| Pumping Station with Rules | PumpRule.p03 |
| Rule Operations | Canal.p01 |
| Simplified Physical Breaching | BaldEDmbrk.p01 |
| UngagedAreaInflows | Ungaged_area.p03, Ungaged_area.p02 |
| Unsteady Flow Encroachment Analysis | UnsteadyEncroachment.p03, UnsteadyEncroachment.p02 |

### Normal Depth

| Example | Plan File |
|---------|-----------|
| Balde Eagle Creek | BaldEagle.p01, BaldEagle.p02 |
| ContractionExpansionMinorLosses | ContExp.p01, ContExp.p02, ContExp.p03, ContExp.p04 |
| Culvert Hydraulics | Beav_Culvert.p01 |
| Dam Breaching | BaldEagleDamBrk.p07, BaldEagleDamBrk.p02, BaldEagleDamBrk.p10, BaldEagleDamBrk.p06, BaldEagleDamBrk.p08, BaldEagleDamBrk.p04, BaldEagleDamBrk.p03, BaldEagleDamBrk.p05, BaldEagleDamBrk.p09, BaldEagleDamBrk.p11 |
| JunctionHydraulics | JunctionHydraulics.p01, JunctionHydraulics.p02, JunctionHydraulics.p04, JunctionHydraulics.p03 |
| Lateral Structure connected to a River Reach | 3Reach_with_Lat.p01 |
| Multiple Reaches with Hydraulic Structures | 3ReachUnsteady.p01, 3ReachUnsteady.p02 |
| Rule Operations | Canal.p01 |
| Simplified Physical Breaching | BaldEDmbrk.p01 |
| UngagedAreaInflows | Ungaged_area.p03, Ungaged_area.p02 |

### Gate Opening

| Example | Plan File |
|---------|-----------|
| Balde Eagle Creek | BaldEagle.p01, BaldEagle.p02 |
| ContractionExpansionMinorLosses | ContExp.p01, ContExp.p02, ContExp.p03, ContExp.p04 |
| Dam Breaching | BaldEagleDamBrk.p07, BaldEagleDamBrk.p02, BaldEagleDamBrk.p10, BaldEagleDamBrk.p06, BaldEagleDamBrk.p08, BaldEagleDamBrk.p04, BaldEagleDamBrk.p03, BaldEagleDamBrk.p05, BaldEagleDamBrk.p09, BaldEagleDamBrk.p11 |
| Elevation Controled Gates | ElevControledGates.p01 |
| Inline Structure with Gated Spillways | Inline_3Gates.p01 |
| Lateral Strcuture with Gates | TwoGateGrps.p01 |
| Lateral Structure with Culverts and Gates | Beav_Lateral.p01 |
| Simplified Physical Breaching | BaldEDmbrk.p01 |

### Stage Hydrograph

| Example | Plan File |
|---------|-----------|
| Internal Stage and Flow Boundary Condition | IBStageFlowTest.p01 |
| Pumping Station with Rules | PumpRule.p03 |
| Rule Operations | Canal.p01 |

### Lateral Inflow Hydrograph

| Example | Plan File |
|---------|-----------|
| Dam Breaching | BaldEagleDamBrk.p07, BaldEagleDamBrk.p02, BaldEagleDamBrk.p10, BaldEagleDamBrk.p06, BaldEagleDamBrk.p08, BaldEagleDamBrk.p04, BaldEagleDamBrk.p03, BaldEagleDamBrk.p05, BaldEagleDamBrk.p09, BaldEagleDamBrk.p11 |
| Multiple Reaches with Hydraulic Structures | 3ReachUnsteady.p01, 3ReachUnsteady.p02 |
| Pumping Station | Pumps.p01 |
| Rule Operations | Canal.p01 |
| UngagedAreaInflows | Ungaged_area.p03, Ungaged_area.p02 |

### Uniform Lateral Inflow Hydrograph

| Example | Plan File |
|---------|-----------|
| Dam Breaching | BaldEagleDamBrk.p07, BaldEagleDamBrk.p02, BaldEagleDamBrk.p10, BaldEagleDamBrk.p06, BaldEagleDamBrk.p08, BaldEagleDamBrk.p04, BaldEagleDamBrk.p03, BaldEagleDamBrk.p05, BaldEagleDamBrk.p09, BaldEagleDamBrk.p11 |
| UngagedAreaInflows | Ungaged_area.p03, Ungaged_area.p02 |

### Unknown or Special Conditions

| Example | Plan File |
|---------|-----------|
| Culverts with Flap Gates | CulvertFlap.p01 |
| Elevation Controled Gates | ElevControledGates.p01 |
| Internal Stage and Flow Boundary Condition | IBStageFlowTest.p01 |
| Lateral Structure Overflow Weir | LatMultXS.p01 |
| Lateral Structure with Culverts | LatWeirCulverts.p01 |
| Levee Breaching | LeveeBreach.p01 |
| Pumping Station with Rules | PumpRule.p03 |
| UngagedAreaInflows | Ungaged_area.p03, Ungaged_area.p02 |
| Unsteady Flow Encroachment Analysis | UnsteadyEncroachment.p03, UnsteadyEncroachment.p02 |

## Category: 2D Unsteady Flow Hydraulics

# HEC-RAS Unsteady Flow Examples Analysis

## Version 6.30

### Example: Muncie

| Boundary Condition Types | Plan Files |
|--------------------------|------------|
| Flow Hydrograph          | Muncie.p01 (Unsteady Multi  9-SA run) |
| Normal Depth             | Muncie.p03 (Unsteady Run with 2D 50ft Grid) |
|                          | Muncie.p04 (Unsteady Run with 2D 50ft User n Value R) |

| Project | Version | Boundary Conditions | Plans |
|---------|---------|---------------------|-------|
| Muncie  | 6.30    | - Flow Hydrograph (White River, Muncie, Station 15696.24) | 1. Muncie.p01: Unsteady Multi 9-SA run |
|         |         | - Normal Depth (White River, Muncie, Station 237.6455)    | 2. Muncie.p03: Unsteady Run with 2D 50ft Grid |
|         |         |                                                           | 3. Muncie.p04: Unsteady Run with 2D 50ft User n Value R |

## Version 6.00

### Example: BaldEagleCrkMulti2D

| Boundary Condition Types | Plan Files |
|--------------------------|------------|
| Flow Hydrograph          | BaldEagleDamBrk.p06 (Gridded Precip - Infiltration) |
| Gate Opening             |  |
| Lateral Inflow Hydrograph |  |
| Uniform Lateral Inflow Hydrograph |  |
| Normal Depth             |  |

| Project | Version | Boundary Conditions | Plans |
|---------|---------|---------------------|-------|
| BaldEagleCrkMulti2D | 6.00 | - Flow Hydrograph | 1. BaldEagleDamBrk.p06: Gridded Precip - Infiltration |
|                     |      | - Gate Opening |  |
|                     |      | - Lateral Inflow Hydrograph |  |
|                     |      | - Uniform Lateral Inflow Hydrograph |  |
|                     |      | - Normal Depth |  |

## Version 5.10

### Example: BaldEagleCrkMulti2D

| Boundary Condition Types | Plan Files |
|--------------------------|------------|
| Flow Hydrograph          | BaldEagleDamBrk.p13 (PMF with Multi 2D Areas) |
| Gate Opening             | BaldEagleDamBrk.p15 (1d-2D Dambreak Refined Grid) |
| Lateral Inflow Hydrograph | BaldEagleDamBrk.p02 (SA to Detailed 2D Breach) |
| Uniform Lateral Inflow Hydrograph | BaldEagleDamBrk.p05 (Single 2D area with Bridges FEQ) |
| Normal Depth             | BaldEagleDamBrk.p01 (SA to Detailed 2D Breach FEQ) |

| Project | Version | Boundary Conditions | Plans |
|---------|---------|---------------------|-------|
| BaldEagleCrkMulti2D | 5.10 | - Flow Hydrograph | 1. BaldEagleDamBrk.p13: PMF with Multi 2D Areas |
|                     |      | - Gate Opening | 2. BaldEagleDamBrk.p15: 1d-2D Dambreak Refined Grid |
|                     |      | - Lateral Inflow Hydrograph | 3. BaldEagleDamBrk.p02: SA to Detailed 2D Breach |
|                     |      | - Uniform Lateral Inflow Hydrograph | 4. BaldEagleDamBrk.p05: Single 2D area with Bridges FEQ |
|                     |      | - Normal Depth | 5. BaldEagleDamBrk.p01: SA to Detailed 2D Breach FEQ |

## Version 5.00

### Example: BaldEagleCrkMulti2D

| Boundary Condition Types | Plan Files |
|--------------------------|------------|
| Flow Hydrograph          | BaldEagleDamBrk.p17 (2D to 1D No Dam) |
| Gate Opening             | BaldEagleDamBrk.p18 (2D to 2D Run) |
| Lateral Inflow Hydrograph | BaldEagleDamBrk.p19 (SA to 2D Dam Break Run) |
| Uniform Lateral Inflow Hydrograph | BaldEagleDamBrk.p03 (Single 2D Area - Internal Dam Structure) |
| Normal Depth             | BaldEagleDamBrk.p04 (SA to 2D Area Conn - 2D Levee Structure) |

| Project | Version | Boundary Conditions | Plans |
|---------|---------|---------------------|-------|
| BaldEagleCrkMulti2D | 5.00 | - Flow Hydrograph | 1. BaldEagleDamBrk.p17: 2D to 1D No Dam |
|                     |      | - Gate Opening | 2. BaldEagleDamBrk.p18: 2D to 2D Run |
|                     |      | - Lateral Inflow Hydrograph | 3. BaldEagleDamBrk.p19: SA to 2D Dam Break Run |
|                     |      | - Uniform Lateral Inflow Hydrograph | 4. BaldEagleDamBrk.p03: Single 2D Area - Internal Dam Structure |
|                     |      | - Normal Depth | 5. BaldEagleDamBrk.p04: SA to 2D Area Conn - 2D Levee Structure |

### Need the same tables for geometry key types

Also need to open and save all example files in 6.5 using HECRASController (making a function to do this)

This can be used to help find examples if the user wants to do something that the library is currently not capable of.  

Using the RasExamples class, the user can extract these files and inspect them to find specific examples.  